[코드숨 5주차 과제] 7월 7일 목요일

Jongveloper opened the pull request:
안녕하세요 트레이너님!
우선은 강의를 계속 돌려보고 있습니다!
기초적인 리덕스를 연결하는 작업만 우선 진행하였고
순차적으로 과제 진행하겠습니다!

Jongveloper committed:
fix 초기 세팅

Jongveloper committed:
test(reducer) setCategories

daadaadaah commented:
좋습니다!👍 일하면서 과제하시느라 힘드실텐대, 코드숨 기간 동안만이라도 바짝 힘내서 공부해봅시당!👊
계속 작업해진행해주세요🚀

Jongveloper commented:
카테고리 테스트를 먼저 작성하고 커밋을 못했네요...!
하지만 expect부분을 검증하지 않고 실수로
reducerHandler를 먼저 작성해서 TDD에서는 벗어났던 것 같습니다!
테스트 코드를 먼저 작성하기는 하지만 검증을 안하면 의미가 없는데...!
이 부분 유의하며 과제 진행하겠습니다!
이번같은 실수는 하지 않도록 하겠습니다!
아직은 익숙치 않아서 실수를 하네요!! ㅠㅠ
P.S 코드숨 기간도 바짝 힘내서 열심히 공부하고 코드숨 기간이 끝나더라도 힘내서 공부를 하겠습니다!ㅎㅎㅎㅎ 감사합니다!

daadaadaah commented:
지금처럼 commit 메시지 빠진거나 수정하고 싶은 거 있을 때, git 을 좀더 공부해보면 좋아요!
과제 빨리 끝내거나 시간 나시면 git commit 메시지 취소, 삭제, 수정 등등 다양한 git 명령어 공부해보세요!
오늘도 열코하세용🔥

Jongveloper committed:
test (setRegions) 테스트 코드 작성

Jongveloper committed:
fix(reducer) setRegions
setRegions 테스트를 통과하였습니다.
테스트가 많아짐에 따라 가독성을 높일 수 있도록 리팩토링하였습니다.

Jongveloper committed:
test (Category) Category 컴포넌트 테스트를 작성하였습니다.

Jongveloper committed:
fix (Category) 카테고리 테스트 통과

Jongveloper committed:
test(Regions) 테스트 코드 작성

Jongveloper committed:
fix(Regions) 테스트 통과

Jongveloper committed:
test(CategoryContainer) 테스트 코드 작성

Jongveloper committed:
refactor(Category) 구조 변경으로 인한 수정

Jongveloper committed:
test(RegionsContainer) 테스트코드 작성

Jongveloper committed:
fix(RegionsContainer) 테스트 통과

Jongveloper commented:
안녕하세요 트레이너님!
ㅎㅎㅎㅎ리뷰요청드립니다!
TDD로 구현을 하려니 아직 익숙치않아 어렵네요! ㅎㅎㅎ
항상 감사합니다!

daadaadaah commented on src/store/reducer.test.js:
```diff
@@ -0,0 +1,19 @@
+describe('setCategories', () => {
+  it('카테고리가 변경됩니다.', () => {
+    const categories = [
+      { id: 1, name: '한식' },
```
저는 테스트 데이터가 배열일 경우, 테스트 데이터를 2~3개정도로 하는 편입니다!

daadaadaah commented on src/components/Category.test.jsx:
```diff
@@ -0,0 +1,19 @@
+  it('카테고리 목록이 보여집니다.', () => {
+    const categories = {
+      id: 1, name: '한식',
+    };
```
1. 객체보다 배열에 어울리는 변수명인 것 같아요!
2. `categories` 라는 테스트용 데이터가 다른 테스트에서도 쓰이는데, 제가 이런 데이터를 어떻게 관리하면 좋다고 했을까요?ㅎㅎ

daadaadaah commented on src/components/Category.jsx:
```diff
@@ -0,0 +1,15 @@
+export default function Category({ categories }) {
```
복수형으로 네이밍하시는게 좀더 적절할 것 같아요!

Jongveloper replied:
아하 ㅎㅎ 감사합니다 리팩토링 진행하겠습니다!

daadaadaah commented on src/store/reducer.test.js:
```diff
@@ -0,0 +1,46 @@
+      );
+      expect(state).toEqual(initailState);
+    });
+  });
+  describe('setCategories', () => {
```
맥락 구분을 위해 한줄 띄어줍시당!

Jongveloper replied:
넵!ㅎㅎ

daadaadaah reviewed (commented):
마음을 좀더 차분하게 가지고 TDD를 진행해보세요! commit 내역을 보면, 테스트 작성 커밋에 (미완성 구현이지만) 구현 코드가 들어가 있어요!(예 : https://github.com/CodeSoom/react-week5-assignment-1/pull/106/commits/38ee60bb889330fefdb5bacd7ae89d0c4e8213b2) 만약, 각 단계별로 어떤 작업을 해야하는지 감이 안오신다면, 언제든 문의주시구용😃 

1. 테스트가 통과 된 후에도 리팩터링 작업(코드에 중복이 없는지 또는 fixture로 관리할 수 있는지)을 진행해보세요! 
2. "작업별"로 commit 메시지 관리해보도록 노력해봅시당! 사소한 작업이라도 분리해서 commit 메시지를 남겨주세용!
3. 지금까지 UI 컴포넌트 -> Container 컴포넌트 작업하시는 순서(=상향식)로 진행하셨는데, 레스토랑 목록은 하향식으로 한번 도전해보세요!

daadaadaah commented on src/container/RegionsContainer.test.jsx:
```diff
@@ -0,0 +1,25 @@
+
+describe('RegionsContainer', () => {
+  useSelector.mockImplementation((selector) => selector({
+    regions: [
+      { id: 1, name: '부산' },
+      { id: 2, name: '대구' },
+    ],
```
`regions`, `categories` 이런 테스트용 데이터가 다른 테스트에서도 쓰이는데, 제가 이런 데이터를 어떻게 관리하면 좋다고 했을까요?ㅎㅎ

Jongveloper replied:
fixture로 빼는 것을 생각했는데 ㅎㅎ 아직은 중복적으로 사용하지 않는데
(나중에 사용할 것을 알지만) 최대한 중복적인 부분을 없애는 경험을 하고싶어
중복이 될 때 fixture로 뺄 생각이었는데 ㅎㅎㅎ...ㅠㅠㅠ
처음부터 뺄 걸 그랬네요! ㅎㅎㅎ
앞으로는 중복될 것 같은 테스트 데이터는 미리 fixture로 빼고 진행하겠습니다!ㅎㅎㅎㅎ
항상 좋은 피드백 감사합니다!

daadaadaah replied:
오호! 생각하고 계셨었군요! 다행이이예요ㅎㅎ
미리 빼는 걸 의도한 건 아닙니당!
제가 코멘트 드린 이유는 
현재 작업 내역 전체를 보면 regions 등의 테스트 데이터가 RegionsContainer.test.jsx에서는 중복되게 사용하고 있지 않지만, 여러 곳(=다른 테스트 파일들)에서 사용하고 있기 때문이예요!
이 점 참고하셔서 PR 날려주세요🚀

Jongveloper replied:
앗 넵 ㅎㅎ 감사합니다!!
참고하여 PR 날리겠습니다!

Jongveloper committed:
refactor (Naming) 명확한 이름으로 변경

Jongveloper committed:
refactor (reducer.test.jsx) 가독성을 위한 리팩토링

Jongveloper committed:
refactor (fixture) 중복 제거

Jongveloper committed:
fix(stroe) redux-thunk 설정

Jongveloper committed:
fix(App) RegionsContainer 와 CategoriesContainer 연결
둘 다 컨테이너에서 불러오기 때문에 공백을 두지 않았습니다.

Jongveloper committed:
refactor(CategoriesContainer) 파일 이름 변경

Jongveloper committed:
fix(categories) 중복 제거를 위한 fixture 생성

Jongveloper committed:
test(App) 테스트 코드 작성

Jongveloper committed:
refactor(categories) 중복 제거

Jongveloper committed:
fix(api) api 함수 정의

Jongveloper committed:
fix(async-actions) async-actions 정의

Jongveloper commented:
안녕하세요 트레이너님!
코드리뷰 요청드립니다!
ㅎㅎ 강의를 여러번 다시 보느라 조금 PR이 늦었습니다!
항상 감사합니다!

daadaadaah commented on src/App.test.jsx:
```diff
@@ -0,0 +1,39 @@
+    ));
+  }
+
+  it('regions가 보여집니다.', () => {
+    const { getByText } = renderApp();
+
+    expect(getByText(regions[0].name)).not.toBeNull();
+  });
+
+  it('categories가 보여집니다.', () => {
+    const { getByText } = renderApp();
+
+    expect(getByText(categories[1].name)).not.toBeNull();
+  });
```
복수형이므로, forEach 사용해서 여러개가 보이는 걸로 테스트해도 좋을 것 같아요!

Jongveloper replied:
앗 그 방법이 있는데 ㅎㅎ 제가 생각이 짧았네요!
항상 좋은 피드백 감사합니다!

daadaadaah reviewed (commented):
아샬님의 강의 여러번 보신거 너무 좋아요!👍 늦더라도 데일리 PR 계속 날려주시는 점도 너무 좋구요! 👏
비동기 액션들에 대한 테스트 추가해주시면서 작업해주시면 될 것 같아요!
날이 더워 체력적으로 힘드실텐대, 지치지 마시구~ 이번주도 같이 화아팅해봐용🔥  
작업별로 Commit 남기시는 거 잊지마시구요!ㅎㅎ다음 PR 기다리고 있을게요🚀

Jongveloper committed:
refactor(App) categories, regions 여러개가 보이는 거로 테스트 수정

Jongveloper committed:
fix(async-actions) 비동기 mocks 추가

Jongveloper committed:
refactor(__mock__) 조금 더 명확하게 구분

Jongveloper committed:
refactor(api) baseURL 범용성있게 변경

Jongveloper committed:
test(api) 테스트 코드 작성

daadaadaah commented:
계속 작업중이신가용?ㅎㅎ 작업 다 완료되시면, 코멘트 남겨주세용😄

Jongveloper committed:
test(RegionsContainer) async 테스트 코드 작성

Jongveloper committed:
refactor(RegionsContainer) async-actions jest.mock추가
이 부분 2cent님 코드를 참고하여 작성하였는데!
async-actions를 mock해야하는 이유를 잘 모르겠습니다.....!

Jongveloper committed:
fix(RegionsContainer) 테스트 통과

Jongveloper committed:
refactor(App.test) regions 테스트 api 호출 함수로 대체

Jongveloper committed:
refactor(CategoriesContainer) Api 호출 테스트 작성

Jongveloper committed:
refactor(CategoriesContainer) mocking

Jongveloper committed:
fix(CategoriesContainer) 테스트 통과

Jongveloper commented:
안녕하세요 트레이너님!
리뷰요청 드립니다!
작업시간이 너무 늦어져 죄송합니다..!
항상 감사합니다!

daadaadaah commented on src/__mocks__/api.js:
```diff
@@ -0,0 +1,6 @@
+import categories from '../fixtures/categories';
+import regions from '../fixtures/regions';
+
+export const fetchRegions = jest.fn(async () => regions);
+
+export const fetchCategories = jest.fn(async () => categories);
```
여기서, 데이터를 고정적으로 넣어주게 되면, 테스트 코드 작성할 때, 어떤 문제가 발생할까요? 🤔 
간단하게 아래처럼 표현하는 것과 차이점에 대해 한번 생각해보세요!
```js
export const fetchRegions = jest.fn();
export const fetchCategories = jest.fn();
```

Jongveloper replied:
앗 한번 어떤 문제가 생길지 생각해보겠습니다!

daadaadaah commented on src/components/Categories.test.jsx:
```diff
@@ -0,0 +1,16 @@
+      />
+    ));
+
+    expect(container).toHaveTextContent('한식');
```
여러개가 보이는 테스트이므로, forEach 사용해서 여러개가 보이는 걸로 테스트해도 좋을 것 같아요!

Jongveloper replied:
헉 이런 실수를...! 리팩토링하겠습니다!

daadaadaah commented on src/components/Regions.test.jsx:
```diff
@@ -0,0 +1,17 @@
+      />
+    ));
+
+    expect(container).toHaveTextContent('대구');
```
여러개가 보이는 테스트이므로, forEach 사용해서 여러개가 보이는 걸로 테스트해도 좋을 것 같아요!

daadaadaah commented on src/container/CategoriesContainer.test.jsx:
```diff
@@ -0,0 +1,47 @@
+describe('CategoriesContainer', () => {
+  const dispatch = jest.fn();
+
+  useDispatch.mockImplementation(() => dispatch);
+  useSelector.mockImplementation((selector) => selector({
```
맥락 구분해주세요!

Jongveloper replied:
헉..리팩토링하겠습니다!

daadaadaah commented on src/container/CategoriesContainer.test.jsx:
```diff
@@ -0,0 +1,47 @@
+    jest.clearAllMocks();
+  });
+
+  it('categories API를 호출합니다.', () => {
```
api 호출 보다 "어떤 액션을 dispatch 한다"라고 표현하는게 더 좋을 것 같아요!

Jongveloper replied:
아아 표현이 애매했네요! 리팩토링하겠습니다!

daadaadaah commented on src/container/CategoriesContainer.test.jsx:
```diff
@@ -0,0 +1,47 @@
+    const { container } = renderCategoriesContainer();
+
+    expect(dispatch).toHaveBeenCalledWith(getCategories());
+    categories.forEach(({ name }) => {
+      expect(container).toHaveTextContent(name);
+    });
```
밑에 41번째 줄의 테스트와 이 테스트로 기대하는 것은 동일한 것 같아요!
둘 중에 하나를 수정해주는게 좋을 것 같아요!

daadaadaah commented on src/container/RegionsContainer.test.jsx:
```diff
@@ -0,0 +1,47 @@
+  const dispatch = jest.fn();
+
+  useDispatch.mockImplementation(() => dispatch);
+  useSelector.mockImplementation((selector) => selector({
```
맥락 구분해주세요!

daadaadaah commented on src/container/RegionsContainer.test.jsx:
```diff
@@ -0,0 +1,47 @@
+
+    expect(container).toHaveTextContent('부산');
+    expect(container).toHaveTextContent('대구');
+  });
```
[CategoriesContainer.test.jsx](https://github.com/CodeSoom/react-week5-assignment-1/pull/106/files#diff-7c8ba870ca216e11eebe1862181aadcbcf1d2d1e0ecc6c86ef1811ea8f7ee8d8)에 드린 코멘트 참고해서 수정해주세요!

daadaadaah commented on src/store/reducer.test.js:
```diff
@@ -0,0 +1,47 @@
+
+  describe('setCategories', () => {
+    it('카테고리리스트가 변경됩니다.', () => {
+      const categories = [
+        { id: 1, name: '한식' },
+      ];
```
fixture에 있는 데이터를 활용해도 좋을 것 같아요!

Jongveloper replied:
앗..이부분을 수정을 못했네요!
좋은 피드백 감사합니다!

daadaadaah commented on src/store/reducer.test.js:
```diff
@@ -0,0 +1,47 @@
+
+  describe('setRegions', () => {
+    it('지역리스트가 변경됩니다.', () => {
+      const regions = [
+        { id: 101, name: '부산' },
+      ];
```
fixture에 있는 데이터를 활용해도 좋을 것 같아요!

Jongveloper replied:
앗..이부분을 수정을 못했네요!
좋은 피드백 감사합니다!

daadaadaah commented on src/services/api.test.js:
```diff
@@ -0,0 +1,27 @@
+
+import { baseURL, fetchCategories, fetchRegions } from './api';
+
+describe('api', () => {
+  describe('fetchRegions', () => {
+    beforeEach(() => {
+      global.fetch = jest.fn(() => Promise.resolve({
+        json: () => Promise.resolve(regions),
+      }));
+    });
+
+    it('지역 목록 Api를 호출합니다.', async () => {
+      await fetchRegions();
+
+      expect(fetch).toHaveBeenCalledWith(`${baseURL}/regions`);
+    });
+  });
+
+  describe('fetchCategories', () => {
+    it('카테고리 목록 Api를 호출합니다.', async () => {
+      await fetchCategories();
+
+      expect(fetch).toHaveBeenCalledWith(`${baseURL}/categories`);
+    });
+  });
+});
```
각 api를 통해 어떤 상황을 확인해보고 싶으신지 한번 생각해보세요!
api가 호출되면 어떤 상황이 벌어질까요? 
```js
describe('fetchCategories', () => {
    it('(fetchCategories 호출하여 기대하는 결과)', () => {
    })
})
```

Jongveloper replied:
앗 호출을 하여 받아오는건데 명확하지 못한 표현이었던 것 같습니다 감사합니다!

daadaadaah replied:
테스트 자체도 수정이 필요합니다!ㅎㅎ

daadaadaah commented on src/services/api.js:
```diff
@@ -0,0 +1,17 @@
+export const baseURL = 'https://eatgo-customer-api.ahastudio.com';
```
export 해주시는 이유가 혹시 테스트 코드에서 사용하기 때문인가용?ㅎㅎ

Jongveloper replied:
앗 맞습니다..!

daadaadaah replied:
api.test.js 수정해보시고, export 계속 필요한지에 대해 한번 생각해보세요!ㅎㅎ

Jongveloper replied:
넵!! ㅎㅎ 전체적으로 리팩토링 할 부분이 있는 것 같습니다! ㅎㅎ
리팩토링진행하겠습니다!
감사합니다!

daadaadaah commented on src/services/api.test.js:
```diff
@@ -0,0 +1,27 @@
+describe('api', () => {
+  describe('fetchRegions', () => {
+    beforeEach(() => {
+      global.fetch = jest.fn(() => Promise.resolve({
+        json: () => Promise.resolve(regions),
+      }));
```
`mockResolvedValue`를 활용하면, 좀더 간단히 표현할 수 있어요!

Jongveloper replied:
앗...! ㅎㅎ 한번 자료 찾아보겠습니다 감사합니다!

daadaadaah reviewed (changes requested):
수고하셨어요! 👏
리뷰어가 드리는 피드포워드에 대해 한번 "왜 이런 피드포워드를 줄까"에 대해 고민해보는 시간을 가져보시는 것도 좋을 것 같아요!ㅎㅎ다음 PR 기대하고 있을게요🚀

Jongveloper committed:
refactor(mocks) 고정적인 값 유동적으로 변경
api response가 변경되면 mock도 수정해야하는 문제가 발생할 것 같습니다!

Jongveloper committed:
refactor(Categories.test) 여러개가 보이는 걸로 변경

Jongveloper committed:
refactor(Regions.test) 여러개가 보이는 테스트로 변경

Jongveloper committed:
refactor(CategoriesContainer.test) 맥락 구분

Jongveloper committed:
refactor(CategoriesContainer) 중복되는 테스트 제거 및 테스트 문구 명확하게 변경

Jongveloper committed:
refactor(RegionsContainer.test) 중복되는 테스트 제거 및 테스트 문구 변경

Jongveloper committed:
refactor(reducer.test) fixture 활용

Jongveloper committed:
refactor(api) mockResolvedValue
이 부분에 대해 좀 이해가 안되는 부분이 있어..혹시 참고할만한 자료 있을까요!?...ㅠㅠ

Jongveloper commented:
안녕하세요 트레이너님! 
mockResolvedValue 부분이 이해가 잘 되지않습니다..
매개변수로 resolve되었을 때 기댓값을 넣어주면 되는걸까요!?

Jongveloper committed:
test(async-actions) 테스트 코드 작성

Jongveloper committed:
test(fetchRestaurants) 테스트코드 작성

Jongveloper committed:
refactor(fetchRestaurants) 매개변수

Jongveloper committed:
fix(fetchRestaaurants) 테스트 통과

Jongveloper committed:
test(getRestaurants) 테스트 코드 작성

Jongveloper committed:
fix(reducer) restaurants 배열 추가

Jongveloper committed:
fix(fixtures) restaurants 추가

Jongveloper committed:
test(setRetaurants) 테스트 추가

Jongveloper committed:
fix(setRestaurants) 테스트 통과

Jongveloper committed:
test(getRestaurants) 테스트 코드 추가

Jongveloper committed:
fix(getRestaurants) 테스트 통과

Jongveloper committed:
test(categories, restaurants) 테스트 추가

Jongveloper committed:
refactor(getRestaurants) 코드 수정

Jongveloper committed:
refactor(getRestaurants.test) 함수변경으로 인한 수정

Jongveloper commented:
안녕하세요 트레이너님!
오늘은 늦어도 너무 늦을 것 같아 따로 DM을 드리지 못했습니다...!
내일 과제 완성할 수 있도록 ㅎㅎㅎ열심히 진행하겠습니다
항상 감사합니다!

Jongveloper committed:
fix(reducer) category, region 추가

Jongveloper committed:
test(reducer) 지역 선택, 카테고리 선택 테스트 추가

Jongveloper committed:
refactor(reducer) 사용해야 하는 함수 import

Jongveloper committed:
fix(reducer) 지역선택, 카테고리 선택 테스트 통과

Jongveloper committed:
test(RestaurantsContainer) 테스트 코드 작성

Jongveloper committed:
fix(RestaurantsContainer) 테스트 통과

Jongveloper committed:
refactor(RestaurantsContainer) 맥락 구분

Jongveloper committed:
test(selectRegion) 테스트 추가

Jongveloper committed:
refactor(RegionsContainer) 중복되는 테스트 제거

Jongveloper committed:
fix(RegionsContainer) 테스트 통과

Jongveloper committed:
test(Regions) 테스트코드 작성

Jongveloper committed:
fix(Regions) 테스트 통과

daadaadaah commented:
작업중이신가요?ㅎㅎ

Jongveloper commented:
> 작업중이신가요?ㅎㅎ

ㅠㅠ오전에 작업하다가 잠시 급한일이 생겨서...!
저녁부터 다시 과제 진행할 예정입니다!!ㅠㅠ

daadaadaah commented:
> > 작업중이신가요?ㅎㅎ
> 
> ㅠㅠ오전에 작업하다가 잠시 급한일이 생겨서...! 저녁부터 다시 과제 진행할 예정입니다!!ㅠㅠ

아핫!ㅎㅎㅎ짬내서 작업을 진행하셨군요! 잘하셨어요👍👍👍
다만, 다음에는 작업이 완료되었으면, 완료되었다는 코멘트 하나만 남겨주시면 좋을 것 같아요!😃

Jongveloper commented:
> > > 작업중이신가요?ㅎㅎ
> 
> > 
> 
> > ㅠㅠ오전에 작업하다가 잠시 급한일이 생겨서...! 저녁부터 다시 과제 진행할 예정입니다!!ㅠㅠ
> 
> 
> 
> 아핫!ㅎㅎㅎ짬내서 작업을 진행하셨군요! 잘하셨어요👍👍👍
> 
> 다만, 다음에는 작업이 완료되었으면, 완료되었다는 코멘트 하나만 남겨주시면 좋을 것 같아요!😃

앗 어제는 너무 늦게 완료될 것 같아ㅠㅠ 따로 dm을 못드렸습니다ㅠㅠ 이번 주에 좀 자주 늦는 것 같아 너무 죄송해서요..!ㅠㅠ
작업 완료되면 꼭 코멘트 남기겠습니다!
항상 감사합니다!

daadaadaah commented on src/App.test.jsx:
```diff
@@ -0,0 +1,66 @@
+    ));
+  }
+
+  it('regions API를 호출합니다.', () => {
+    const { container } = renderApp();
+
+    expect(dispatch).toHaveBeenCalledWith(getRegions());
+
+    regions.forEach(({ name }) => {
+      expect(container).toHaveTextContent(name);
+    });
+  });
+
+  it('categories API를 호출합니다.', () => {
+    const { container } = renderApp();
+
+    expect(dispatch).toHaveBeenLastCalledWith(getCategories());
+
+    categories.forEach(({ name }) => {
+      expect(container).toHaveTextContent(name);
+    });
+  });
+
+  it('restaurants API를 호출합니다.', () => {
+    const { container } = renderApp();
+
+    expect(dispatch).toHaveBeenCalledWith(getRestaurants());
+
+    restaurants.forEach(({ name }) => {
+      expect(container).toHaveTextContent(name);
+    });
+  });
```
API 호출은 어디서 하는지 생각해보시고, App 입장에서 무엇을 하는지를 테스트에 표현해주세요!

daadaadaah commented on src/container/CategoriesContainer.test.jsx:
```diff
@@ -0,0 +1,42 @@
+import { getCategories } from '../store/async-actions';
+
+jest.mock('react-redux');
+jest.mock('../store/async-actions');
```
비동기 액션을 mock 해주시는 이유가 무엇인가용?ㅎㅎ

Jongveloper replied:
비동기 액션을 mock하지 않으면 테스트가 깨지더라구요!ㅠㅠ
혹시 다른 방법이 있을까요!?

daadaadaah replied:
`__mocks__` 폴더에 `async-action.js` 파일 하나 만들어서 mocking이 필요한 것들 추가해주시면 좋을 거 같아요!

daadaadaah replied:
이렇게 mocking 해주실 때, `api`랑 `redux-react mocking`한 것처럼 `__mocks__`폴더에 `async-action.js` 하나 만들어서 mocking하는 함수 넣어주시고, 테스트 코드에서 mockImplement 해주셔야됩니당!ㅎㅎ

daadaadaah commented on src/services/api.test.js:
```diff
@@ -0,0 +1,33 @@
+  baseURL, fetchCategories, fetchRegions, fetchRestaurants,
+} from './api';
+
+describe('api', () => {
+  describe('fetchRegions', () => {
+    beforeEach(() => {
+      global.fetch = jest.fn().mockResolvedValue({ json: () => {} });
+    });
+
+    it('지역 목록 Api를 호출합니다.', async () => {
+      await fetchRegions();
+
+      expect(fetch).toHaveBeenCalledWith(`${baseURL}/regions`);
+    });
+  });
+
+  describe('fetchCategories', () => {
+    it('카테고리 목록 Api를 호출합니다.', async () => {
+      await fetchCategories();
+
+      expect(fetch).toHaveBeenCalledWith(`${baseURL}/regions`);
+    });
+  });
+
+  describe('fetchRestaurants', () => {
+    it('레스토랑 목록 Api를 호출합니다.', async () => {
+      await fetchRestaurants({ region: '서울', categoryId: '3' });
+
+      expect(fetch).toHaveBeenCalledWith(`${baseURL}/restaurants?region=서울&category=3`);
+    });
+  });
+});
```
아직 수정 작업이 안되신거죵?ㅎㅎ
https://github.com/CodeSoom/react-week5-assignment-1/pull/106#discussion_r916614995
코멘트 참고해서 수정해주세요~

Jongveloper replied:
엇..분명 바꿨는데..ㅠㅠ 제가 바꿨다 생각하고 변경을 안했나봅니다!
변경하겠습니다!

daadaadaah commented on src/store/async-actions.test.js:
```diff
@@ -0,0 +1,42 @@
+
+const mockStore = configureStore([thunk]);
+
+describe('async-actions', () => {
+  const store = mockStore();
+
+  beforeEach(() => {
+    store.clearActions();
+  });
+
+  describe('getRegions', () => {
+    it('fetchRegions', async () => {
+      await store.dispatch(getRegions());
+
+      expect(fetchRegions).toBeCalled();
+    });
+  });
+
+  describe('getCategories', () => {
+    it('fetchCategories', async () => {
+      await store.dispatch(getCategories());
+
+      expect(fetchCategories).toBeCalled();
+    });
+  });
+
+  describe('getRestaurants', () => {
+    it('fetchRestaurants', async () => {
+      await store.dispatch(getRestaurants({ region: '서울', categoryId: '1' }));
+
+      expect(fetchRestaurants).toBeCalled();
+    });
+  });
+});
```
`redux-mock-store`를 활용하면, dispatch 시에 어떤 action이 호출되었는지까지 확인할 수 있어서, 좀더 의도를 드러낼 수 있어요~ 한번 활용해봅시당!ㅎㅎ

Jongveloper replied:
앗 넵 자료 찾아보고 활용해보겠습니다!

daadaadaah reviewed (changes requested):
매일 조금씩 PR 올려주시는거 너무 잘하고 계세요!ㅎㅎ
남은 주말에도 힘내서 과제 완성시켜 봅시당👊

Jongveloper committed:
refactor(api.test) 명확한 표현으로 변경

Jongveloper committed:
test(CategoriesContainer) selectCategory 테스트 추가

Jongveloper committed:
refactor(selectCategory) 오탈자 수정

Jongveloper committed:
fix(CategoriesContainer) handleClickCategory 구현

Jongveloper committed:
test(Categories) handleClickCategory 테스트 추가

Jongveloper committed:
refactor(handleClickCategories) 테스트 깨지던 오류 수정
regions와 같은 형식의 테스트가 왜 카테고리에서 깨지는 지 이해가 안되는데...ㅠㅠ
혹시 어떤 부분 때문에 테스트가 깨지는건지 알 수 있을까요?

Jongveloper committed:
fix(Categories) 테스트 통과

Jongveloper committed:
refactor(Regions) RORO패턴 적용

Jongveloper committed:
refactor(handleClickCategory) 매개변수 더욱 명확하게 변경

Jongveloper committed:
refactor(restaurants_test) response 값 변경으로 인한 수정

Jongveloper committed:
refactor(async-actions) getActions 활용

Jongveloper commented:
안녕하세요 트레이너님!
트레이너님이 피드백 주신 것 반영하여 리팩토링하고 과제 완료하였습니다!
항상 좋은 피드백 주셔서 과제를 완성할 수 있었던 것 같습니다!
항상 감사합니다!
getActions 부분 자료를 찾아보았는데 명확하게 설명하는 자료를 못찾아서...
혹시 실례가 안된다면 이 부분에 대해 참고자료 부탁드려도 될까요..?!

daadaadaah commented:
> getActions 부분 자료를 찾아보았는데 명확하게 설명하는 자료를 못찾아서...
혹시 실례가 안된다면 이 부분에 대해 참고자료 부탁드려도 될까요..?!

getActions의 어떤 자료가 필요하신거예요??ㅎㅎ

daadaadaah commented on src/components/Regions.jsx:
```diff
@@ -2,11 +2,11 @@ export default function Regions({ regions, selectedRegion, onClick }) {
-            <button type="button" onClick={() => onClick(region.name)}>
-              {region.name}
-              {selectedRegion === region.name && '(V)'}
+        regions.map(({ id, name }) => (
```
이건 객체 구조 분해 입니다! 
RORO 패턴에 대해 다시 공부해보세요!

Jongveloper replied:
앗 제가 객체 구조 분해를 RORO로 혼동하고 있었던 것 같습니다.
RORO패턴에 대해 다시 공부하겠습니다!
좋은 피드백 감사합니다!

daadaadaah commented on src/services/api.test.js:
```diff
@@ -0,0 +1,33 @@
+  });
+
+  describe('fetchCategories', () => {
+    it('getRegions api주소와 함께 fetch됩니다.', async () => {
```
우리가 `fetchCategories`로 기대하는 것은 무엇인지에 대해 다시 생각해보세요!
예를 들어, 주소를 가져오는 `getAddress`라는 함수를 테스트 하고 싶을 때, it문을 "return address" 이렇게 작성할 것 같아요! 테스트도 마찬가지로 getAddrees 결과, 실제 return 값이 address인지 확인할 것 같구요!ㅎㅎ

```js
const address = "서울";
const result = await getAddress();
expect(result).tobe(address);

```

Jongveloper replied:
앗 피드백 감사합니다!
피드백 참고하여 리팩토링 진행하겠습니다!

daadaadaah commented on src/components/Categories.test.jsx:
```diff
@@ -0,0 +1,43 @@
+import Categories from './Categories';
+
+describe('Categories', () => {
+  const defaultCategory = categories[0].id;
```
> 5397e71f70152386c5fea3a6c595a829b708b9e2 질문에 대한 답변

region과 다르게 id를 사용해서 그런 것 같아요! 이런 문제가 똑같이 발생하기 위해서는 변수명을 어떻게하는게 좋을까요?
`const defaultRegion = regions[0].name;`

그리고, 다음에는 테스트가 깨질 때나 오류날 때, 스샷 같은거 같이 첨부해주셨으면 좋겠습니당!ㅎㅎ

Jongveloper replied:
아아 네네!
ㅎㅎㅎ 좋은 피드백 감사합니다
테스트가 깨질 때 스샷을 같이 첨부하겠습니다 감사합니다!

daadaadaah reviewed (changes requested):
> getActions 부분 자료를 찾아보았는데 명확하게 설명하는 자료를 못찾아서...
혹시 실례가 안된다면 이 부분에 대해 참고자료 부탁드려도 될까요..?!

getActions의 어떤 자료가 필요하신거예요??ㅎㅎ redux-mock-store 자료 말씀하신 건가용?ㅎㅎ
https://github.com/reduxjs/redux-mock-store

daadaadaah commented on src/container/RestaurantsContainer.test.jsx:
```diff
@@ -0,0 +1,44 @@
+jest.mock('../store/async-actions');
+
+describe('RestaurantsContainer', () => {
+  const selectedRegion = regions[0].name;
+  const selectedCategory = categories[0].id;
```
통일성 있게 관리해주시는게 좋을 거 같아요! 어디서는 name으로 관리하고 어디서는 id로 관리하면 헷갈릴 것 같아서요!

Jongveloper replied:
아하!! 참고하여 리팩토링 하겠습니다 좋은 피드백 감사합니다!

Jongveloper committed:
refactor(api.test) 테스트를 더욱 명확하게 변경

Jongveloper committed:
refactor(api) test변경으로 인한 baseUrl 범용성이 필요없기에 export 제거

Jongveloper committed:
test(RestaurantsContainer.test) 헷갈릴 수 있는 변수 수정

Jongveloper committed:
refactor(Regions.test) 헷갈릴 수 있는 변수 수정

Jongveloper committed:
refactor(Categories.test) 헷갈릴 수 있는 변수명 변경
트레이너님의 말씀대로 헷갈리게끔 변수를 작성한게 테스트가
깨지게한 원인인 것 같습니다...
이미 변수에서 카테고리의 1번 째 id로 규정을 지었는데
거기서 그걸 인지 못한채로 name을 불러오려하니..당연히 오류가 날 수 밖에 없었습니다.
항상 좋은 피드백 정말 감사합니다!

Jongveloper commented:
안녕하세요 트레이너님!
트레이너님이 주신 피드백 반영하여 과제 완성하였습니다!
트레이너님 덕분에 과제 완성할 수 있었던 것 같습니다
항상 감사합니다!

daadaadaah reviewed (approved):
이번주도 수고하셨어요!🔥
